Put TORCH_LIBRARY in torch/library.h; add custom class API

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36800 Port xnnpack operators to new registration API
* **#36742 Put TORCH_LIBRARY in torch/library.h; add custom class API**

Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

ezyang committed:
Add custom class API to TORCH_LIBRARY API.

Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom_class_detail.h/custom_class.h were moved to ATen/core, there are
some modifications in custom_class.h: the definition of Library::class_
is in the bottom of that header because I need all of the class_
constructors available, but there is a circular dependency between the
two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

jamesr66a commented on torch/custom_class.h:
```diff
@@ -1,273 +1,2 @@
-} // namespace jit
-
-} // namespace torch
+#include <ATen/core/op_registration/custom_class.h>
```
Do we really need to move all this? `custom_class.h` was meant to be easily accessible and readable for libtorch users and I don't like the indirection this is introducing

ezyang replied:
I'm not sure. The operant question is whether or not I can include `torch/` base directory from `ATen/core`. I can *probably* make it work if I push on the build system enough, but my default assumption is it will require a bit of finagling. (Maybe less finagling since custom_class.h already works on mobile, though?)

bhosmer replied:
Should we ask if the same accessibility concerns apply to the new API as to `custom_class.h`? Maybe this is a symptom of the new stuff being too buried.

ezyang committed:
Update on "Add custom class API to TORCH_LIBRARY API."


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom_class_detail.h/custom_class.h were moved to ATen/core, there are
some modifications in custom_class.h: the definition of Library::class_
is in the bottom of that header because I need all of the class_
constructors available, but there is a circular dependency between the
two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add custom class API to TORCH_LIBRARY API."


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom_class_detail.h/custom_class.h were moved to ATen/core, there are
some modifications in custom_class.h: the definition of Library::class_
is in the bottom of that header because I need all of the class_
constructors available, but there is a circular dependency between the
two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add custom class API to TORCH_LIBRARY API."


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom_class_detail.h/custom_class.h were moved to ATen/core, there are
some modifications in custom_class.h: the definition of Library::class_
is in the bottom of that header because I need all of the class_
constructors available, but there is a circular dependency between the
two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add custom class API to TORCH_LIBRARY API."


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom_class_detail.h/custom_class.h were moved to ATen/core, there are
some modifications in custom_class.h: the definition of Library::class_
is in the bottom of that header because I need all of the class_
constructors available, but there is a circular dependency between the
two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add custom class API to TORCH_LIBRARY API."


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom_class_detail.h/custom_class.h were moved to ATen/core, there are
some modifications in custom_class.h: the definition of Library::class_
is in the bottom of that header because I need all of the class_
constructors available, but there is a circular dependency between the
two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang commented:
@jamesr66a OK, PR is updated so that `TORCH_LIBRARY` is now next to `custom_class.h` in torch folder

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

ezyang committed:
Update on "Put TORCH_LIBRARY in torch/library.h; add custom class API"


Now, you can define a custom class inside a TORCH_LIBRARY block.
It looks very similar to what you did before.  Instead of

```
static auto m = torch::class_<Class>("Namespace", "Class").def("foo", foo);
```

you write

```
TORCH_LIBRARY(Namespace, m) {
  m.class_<Class>("Class")
    .def("foo", foo);
}
```

All the old usages still work, but at some point we should start
updating the tutorials when we're ready to go 100% live with the
new pybind11 style API.

custom class API previously lived in torch/ folder and in torch
namespace, so for consistency, the new TORCH_LIBRARY also got
moved to torch/library.h The definition of Library::class_ is in the
bottom of that header because I need all of the class_ constructors
available, but there is a circular dependency between the two headers.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21089648](https://our.internmc.facebook.com/intern/diff/D21089648)

[ghstack-poisoned]

jamesr66a commented:
Reverting due to master-only breakage of pytorch_linux_xenial_py3_clang5_mobile_code_analysis (see https://app.circleci.com/jobs/github/pytorch/pytorch/5207707)

To Run:
```
Construct your PR as usual. Then, on a diff stacked on top of your PR, make the following modifications:

* Edit .circleci/cimodel/data/pytorch_build_data.py: Find the relevant build definition in CONFIG_TREE_DATA and replace the function call to X with XImportant
* Edit .circleci/scripts/should_run_job.sh: Add “exit 0” to the top of this file
* Run .circleci/regenerate.sh

Push this PR. Your requested job will now run on the PR.
```
